File: archive/navvis-cleanroom-alignment/code/src/navvis_recon/cloud_builder.py
"""Reference reconstruction of the non-SLAM laser cloud builder.

Recovered evidence identifies an ordered chain containing NonFiniteXYZFilter,
IntensityNormalizer, IntensityRegionFilter, MultilayerFringeFilter,
RegionFilter, RingRangeFilter and PlaneFilter.  The implementation below keeps
that order and exposes the device-specific geometry as configuration instead of
embedding proprietary G11 calibration numbers.
"""

from __future__ import annotations

from bisect import bisect_right
from dataclasses import dataclass, field
from typing import Callable, Iterable, Sequence

import numpy as np

from .models import LaserPoint, Pose, unit


@dataclass(slots=True)
class AxisAlignedRegion:
    minimum: np.ndarray
    maximum: np.ndarray

    def contains(self, p: np.ndarray) -> bool:
        return bool(np.all(p >= self.minimum) and np.all(p <= self.maximum))


@dataclass(slots=True)
class BuilderConfig:
    min_range_by_ring: dict[int, float] = field(default_factory=dict)
    max_range_by_ring: dict[int, float] = field(default_factory=dict)
    intensity_floor: float = 0.05
    normalize_intensity: bool = True
    unskew: bool = True
    remove_regions: list[AxisAlignedRegion] = field(default_factory=list)
    fringe_regions: list[AxisAlignedRegion] = field(default_factory=list)
    no_motion_translation: float = 0.005
    no_motion_angle_deg: float = 0.05
    plane_distance_threshold: float | None = None


def pose_at(trajectory: Sequence[Pose], timestamp: float) -> Pose:
    if not trajectory:
        raise ValueError("trajectory is empty")
    times = [p.timestamp for p in trajectory]
    i = bisect_right(times, timestamp)
    if i == 0:
        return trajectory[0]
    if i == len(trajectory):
        return trajectory[-1]
    return Pose.interpolate(trajectory[i - 1], trajectory[i], timestamp)


def estimate_ordered_normals(rows: Sequence[Sequence[LaserPoint]]) -> None:
    """Estimate normals from adjacent laser/ring samples in an ordered cloud."""
    for r in range(1, len(rows) - 1):
        width = min(len(rows[r - 1]), len(rows[r]), len(rows[r + 1]))
        for c in range(1, width - 1):
            horizontal = rows[r][c + 1].xyz - rows[r][c - 1].xyz
            vertical = rows[r + 1][c].xyz - rows[r - 1][c].xyz
            n = unit(np.cross(horizontal, vertical))
            ray = rows[r][c].xyz - rows[r][c].origin
            if np.dot(n, ray) > 0:
                n = -n
            rows[r][c].normal = n


def _valid_range(p: LaserPoint, cfg: BuilderConfig) -> bool:
    distance = float(np.linalg.norm(p.xyz - p.origin))
    return cfg.min_range_by_ring.get(p.ring, 0.0) <= distance <= cfg.max_range_by_ring.get(p.ring, np.inf)


def _normalize_intensity(points: list[LaserPoint]) -> None:
    by_ring: dict[int, list[LaserPoint]] = {}
    for p in points:
        by_ring.setdefault(p.ring, []).append(p)
    for ring_points in by_ring.values():
        values = np.array([p.intensity for p in ring_points], dtype=float)
        lo, hi = np.percentile(values, [5, 95]) if len(values) > 1 else (0.0, max(values[0], 1.0))
        span = max(float(hi - lo), 1e-6)
        for p in ring_points:
            p.intensity = float(np.clip((p.intensity - lo) / span, 0, 1))


def _plane_inliers(points: list[LaserPoint], threshold: float, iterations: int = 96) -> set[int]:
    """Small RANSAC plane model matching the PCL model proven in the ELF."""
    if len(points) < 3:
        return set()
    rng = np.random.default_rng(0)
    xyz = np.array([p.xyz for p in points])
    best: np.ndarray = np.array([], dtype=int)
    for _ in range(iterations):
        ids = rng.choice(len(points), 3, replace=False)
        a, b, c = xyz[ids]
        n = unit(np.cross(b - a, c - a))
        if np.linalg.norm(n) < 1e-9:
            continue
        inliers = np.flatnonzero(np.abs((xyz - a) @ n) <= threshold)
        if len(inliers) > len(best):
            best = inliers
    return set(map(int, best))


def build_cloud(
    scans: Iterable[Sequence[LaserPoint]],
    trajectory: Sequence[Pose],
    cfg: BuilderConfig,
    sensor_to_rig: Callable[[LaserPoint], np.ndarray] | None = None,
) -> list[LaserPoint]:
    """Filter, unskew and transform raw points into the map frame.

    SLAM is deliberately absent: ``trajectory`` is an already-estimated input.
    """
    accepted: list[LaserPoint] = []
    for scan in scans:
        for raw in scan:
            if not np.all(np.isfinite(raw.xyz)) or not _valid_range(raw, cfg):
                continue
            p_rig = sensor_to_rig(raw) if sensor_to_rig else raw.xyz
            if any(region.contains(p_rig) for region in cfg.remove_regions):
                continue
            if raw.intensity < cfg.intensity_floor:
                continue
            if any(region.contains(p_rig) for region in cfg.fringe_regions):
                continue
            pose = pose_at(trajectory, raw.timestamp) if cfg.unskew else pose_at(trajectory, scan[0].timestamp)
            accepted.append(
                LaserPoint(
                    xyz=pose.apply(p_rig),
                    timestamp=raw.timestamp,
                    intensity=raw.intensity,
                    ring=raw.ring,
                    origin=pose.translation.copy(),
                    normal=None if raw.normal is None else pose.rotation @ raw.normal,
                    ray_weight=raw.ray_weight,
                )
            )
    if cfg.normalize_intensity and accepted:
        _normalize_intensity(accepted)
    if cfg.plane_distance_threshold:
        # The binary constructs PCL RandomSampleConsensus<ModelPlane>.  Keeping
        # the inliers is useful for calibration/ground extraction modes.
        inliers = _plane_inliers(accepted, cfg.plane_distance_threshold)
        accepted = [p for i, p in enumerate(accepted) if i in inliers]
    return accepted

File: archive/navvis-cleanroom-alignment/code/cpp/src/cloud_builder.cpp
#include "navvis_recon/cloud_builder.hpp"

#include <Eigen/Geometry>

#include <algorithm>
#include <cmath>
#include <numeric>
#include <random>
#include <stdexcept>

namespace navvis_recon {

bool NonFiniteXYZFilter::keep(const LaserPoint& point) const {
    return point.xyz.array().isFinite().all();
}

bool RingRangeFilter::keep(const LaserPoint& point) const {
    const float distance = (point.xyz - point.origin).norm();
    const auto minimum = minimum_range.find(point.ring);
    const auto maximum = maximum_range.find(point.ring);
    const float lower = minimum == minimum_range.end() ? 0.0F : minimum->second;
    const float upper = maximum == maximum_range.end()
                            ? std::numeric_limits<float>::infinity()
                            : maximum->second;
    return distance >= lower && distance <= upper;
}

bool RegionFilter::keep(const LaserPoint& point) const {
    return std::none_of(rejected_regions.begin(), rejected_regions.end(),
                        [&](const AxisAlignedRegion& region) { return region.contains(point.xyz); });
}

bool IntensityRegionFilter::keep(const LaserPoint& point) const {
    if (point.intensity >= minimum_intensity) {
        return true;
    }
    return std::none_of(reflection_regions.begin(), reflection_regions.end(),
                        [&](const AxisAlignedRegion& region) { return region.contains(point.xyz); });
}

bool MultilayerFringeFilter::keep(const LaserPoint& point) const {
    return std::none_of(fringe_regions.begin(), fringe_regions.end(),
                        [&](const AxisAlignedRegion& region) { return region.contains(point.xyz); });
}

std::vector<std::size_t> PlaneFilter::largestPlane(const std::vector<LaserPoint>& cloud) const {
    if (cloud.size() < 3U) {
        return {};
    }
    std::mt19937 generator(0);
    std::uniform_int_distribution<std::size_t> sample(0, cloud.size() - 1);
    std::vector<std::size_t> best;
    for (int iteration = 0; iteration < iterations; ++iteration) {
        const auto ia = sample(generator);
        const auto ib = sample(generator);
        const auto ic = sample(generator);
        if (ia == ib || ia == ic || ib == ic) {
            continue;
        }
        const Vec3f normal = (cloud[ib].xyz - cloud[ia].xyz).cross(cloud[ic].xyz - cloud[ia].xyz);
        if (normal.squaredNorm() < 1.0e-10F) {
            continue;
        }
        const Vec3f unit_normal = normal.normalized();
        std::vector<std::size_t> inliers;
        for (std::size_t i = 0; i < cloud.size(); ++i) {
            if (std::abs(unit_normal.dot(cloud[i].xyz - cloud[ia].xyz)) <= distance_threshold) {
                inliers.push_back(i);
            }
        }
        if (inliers.size() > best.size()) {
            best = std::move(inliers);
        }
    }
    return best;
}

void IntensityNormalizer::normalizePerRing(std::vector<LaserPoint>& cloud) {
    std::unordered_map<std::uint16_t, std::vector<std::size_t>> by_ring;
    for (std::size_t i = 0; i < cloud.size(); ++i) {
        by_ring[cloud[i].ring].push_back(i);
    }
    for (const auto& entry : by_ring) {
        std::vector<float> values;
        values.reserve(entry.second.size());
        for (const auto index : entry.second) {
            values.push_back(cloud[index].intensity);
        }
        std::sort(values.begin(), values.end());
        const auto percentile = [&](float fraction) {
            return values[static_cast<std::size_t>(fraction * static_cast<float>(values.size() - 1))];
        };
        const float low = percentile(0.05F);
        const float high = percentile(0.95F);
        const float span = std::max(high - low, 1.0e-6F);
        for (const auto index : entry.second) {
            cloud[index].intensity = std::clamp((cloud[index].intensity - low) / span, 0.0F, 1.0F);
        }
    }
}

CloudBuilder::CloudBuilder(CloudBuilderOptions options) : options_(std::move(options)) {}

Pose CloudBuilder::poseAt(const std::vector<Pose>& trajectory, double timestamp) {
    if (trajectory.empty()) {
        throw std::invalid_argument("trajectory is empty");
    }
    const auto upper = std::upper_bound(
        trajectory.begin(), trajectory.end(), timestamp,
        [](double time, const Pose& pose) { return time < pose.timestamp; });
    if (upper == trajectory.begin()) {
        return trajectory.front();
    }
    if (upper == trajectory.end()) {
        return trajectory.back();
    }
    return Pose::interpolate(*std::prev(upper), *upper, timestamp);
}

bool CloudBuilder::scanMoved(const Pose& previous, const Pose& current) const {
    const float distance = (current.translation - previous.translation).norm();
    const float angle = Eigen::AngleAxisf(previous.rotation.conjugate() * current.rotation).angle() *
                        180.0F / static_cast<float>(M_PI);
    return distance >= options_.no_motion_translation_m || angle >= options_.no_motion_angle_deg;
}

std::vector<LaserPoint> CloudBuilder::build(
    const std::vector<std::vector<LaserPoint>>& scans,
    const std::vector<Pose>& trajectory,
    const std::function<Vec3f(const LaserPoint&)>& sensor_to_rig) const {
    std::vector<LaserPoint> result;
    std::optional<Pose> previous_scan_pose;
    for (const auto& scan : scans) {
        if (scan.empty()) {
            continue;
        }
        const Pose reference_pose = poseAt(trajectory, scan.front().timestamp);
        if (options_.remove_no_motion_scans && previous_scan_pose &&
            !scanMoved(*previous_scan_pose, reference_pose)) {
            continue;
        }
        previous_scan_pose = reference_pose;
        for (const LaserPoint& raw : scan) {
            LaserPoint rig_point = raw;
            rig_point.xyz = sensor_to_rig ? sensor_to_rig(raw) : raw.xyz;
            bool accepted = true;
            for (const auto& filter : options_.filters) {
                if (!filter->keep(rig_point)) {
                    accepted = false;
                    break;
                }
            }
            if (!accepted) {
                continue;
            }
            const Pose pose = options_.unskew ? poseAt(trajectory, raw.timestamp) : reference_pose;
            rig_point.xyz = pose.apply(rig_point.xyz);
            rig_point.origin = pose.translation;
            if (rig_point.has_normal) {
                rig_point.normal = pose.rotation * rig_point.normal;
            }
            result.push_back(rig_point);
        }
    }
    if (options_.normalize_intensity && !result.empty()) {
        IntensityNormalizer::normalizePerRing(result);
    }
    return result;
}

void CloudBuilder::computeNormalsFromOrderedCloud(std::vector<std::vector<LaserPoint>>& rows) {
    if (rows.size() < 3U) {
        return;
    }
    for (std::size_t row = 1; row + 1 < rows.size(); ++row) {
        const std::size_t width = std::min({rows[row - 1].size(), rows[row].size(), rows[row + 1].size()});
        for (std::size_t column = 1; column + 1 < width; ++column) {
            const Vec3f horizontal = rows[row][column + 1].xyz - rows[row][column - 1].xyz;
            const Vec3f vertical = rows[row + 1][column].xyz - rows[row - 1][column].xyz;
            Vec3f normal = normalizedOr(horizontal.cross(vertical));
            if (normal.dot(rows[row][column].xyz - rows[row][column].origin) > 0.0F) {
                normal = -normal;
            }
            rows[row][column].normal = normal;
            rows[row][column].has_normal = normal.squaredNorm() > 0.5F;
        }
    }
}

}  // namespace navvis_recon
